EIPS/eip-3102.md

```diff
@@ -0,0 +1,189 @@
+
+Additionally, binary trees are expected to be widely used in Eth2, so this path improves forward-compatibility and reduces long-run total complexity for the protocol.
+
+### Path length instead of bit prefix
+
+In order to remove the complexity associated with byte manipulation, only the bit-length of the extension is used to merkelize a node with a prefix.
+
+Storing the length of the path from the root node instead of that from the parent node has the nice property that siblings need not be hashed when deleting a leaf.
+
+![Sibling deletion diagram](../assets/eip-3102/sibling.svg)
+
+_On the left, a trie with the prefix length, and on the right, a trie with the full path length. Each both have values `10000100` and `10000000`. Deleting `10000100` result in the same trie. In the case of the left trie, the sibling node has to be updated, while in the case of the right trie, it need not be._
+
+### Value hashing
+
+Except for the code, all values in the trie are less than 32 bytes. EIP-2926 introduces code chunks, with `CHUNK_SIZE = 32 bytes`. The hashing of the leaf's value could therefore be saved. The authors of the EIP are however considering a future increase of `CHUNK_SIZE`, making `hash(value)` the future-proof choice.
+
+## Backwards Compatibility
+
+A hard fork is required in order for blocks to have a trie root using a different structure.
+
+## Test Cases
+
+This section is still incomplete and will be fleshed out as clients start implementing the support for binary tries.
+
+An initial implementation has been made for Geth. Unit tests can be found [here](https://github.com/gballet/go-ethereum/blob/binary-trie-m5-full-path/trie/binary_test.go).
+
+## Implementation
+
+ * As of [commit 0db87e187dc0bfb96046a47e3d6768c93a2e3331](https://github.com/gballet/multiproof-rs/commit/6d22b1aef9548581826b3c04b3e00d6cc709388c), [multiproof-rs](https://github.com/gballet/multiproof-rs) implements this merkelization rule in the `hash_m5()` function, found in file `src/binary_tree.rs`.
+ * An implementation of this structure for go-ethereum is available [in this branch](https://github.com/gballet/go-ethereum/tree/binary-trie-m5).
+
+## Security Considerations
+
+Issues could arise when performing the transition. In particular, a heavy conversion process could incentivize clients to wait the transition out. This could lead to a lowered network security at the time of the transition. A transition process has been proposed with [EIP-2584](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2584.md).
+
+## Acknowledgment
+
+This EIP has benefited from the valuable input of @poemm, @BoltonBailey, @s1na, @ledgerwatch and @holiman.
+
```

Comment:
```suggestion
```
EIPs are public domain, which means attribution is not required and we want to avoid EIPs turning into a place where people give out props/kudos/etc.  Social media, blogs, discussion forums, personal/company websites, etc. are a much better place for such things than a technical specification.